README
NAME
    Soar::Production::Parser - Parse Soar productions

VERSION
    version 1.122860

SYNOPSIS
      use Soar::Production::Parser;
      use Data::Dumper;
      my $parser = Soar::Production::Parser->new;
      my @trees=$parser->parse_file("foo.soar");
      print Dumper(\@trees);

DESCRIPTION
    Soar is a cognitive modeling architecture for programming and
    experimenting with intelligent agents. Soar is programmed using
    productions that look like this:

            sp{name
                    (state <s>)
                    -->
                    (<s> ^foo bar)
            }

    The preceding production matches any state and adds an element named
    "foo" with the value "bar" to it. Productions can get much more
    complicated than that. This module can be used to parse these
    productions. Underlyingly, a Parse::RecDescent grammar is used to
    convert a production into a parse tree. There are also methods for
    extracting all of the productions from a file string, and to remove
    comments (not that I think you'll ever want to do that!).

NAME
    Soar::Production::Parser - Perl extension for parsing angst grammar
    files

METHODS
  "new"
    Creates a new parser.

  "parse_text"
    Argument: the text of a single Soar production. Returns: a parse tree
    for the given production.

  "get_parses"
    Argument: Reference to array containing text for individual productions.
    Return: Reference to an array containing parse trees for each of the
    productions in the input array reference.

  "no_comment"
    Argument: Text which contains Soar productions or commands Return: Same
    text, but with all comments removed. Comments are indicated with a #
    (pound), optionally preceded by a ; (semicolon) and whitespace.

  "productions"
    This method extracts productions from a given text. It returns a
    reference to an array containing either the text of each of the
    productions, or a parse tree for each of them. Note that all comments
    are removed as a preprocessing step to detecting and extracting
    productions. It takes a set of named arguments: 'file'- the name of a
    file to read. 'text'- the text to split. 'parse'- set to true if the
    return value should be an array of parse trees for the extracted
    productions; otherwise an array containing the production text will be
    returned. For example, if you would like to extract all of the
    productions from a file and print their parse trees, you could do this:

        use Soar::Production::Parser;
            use Data::Dumper;
        
            my $file = shift;
            my $parser = Soar::Production::Parser->new();
            my $parses = $parser->productions(
                    file => $file,
                    parse => 1
            );
        
            for my $prod(@$productions){
                    print Dumper($prod);
            }

SEE ALSO
    The documentation for Soar is located at
    <https://code.google.com/p/soar/>. You may also be interested in what a
    production system is, since this module parses Soar productions:
    <http://en.wikipedia.org/wiki/Production_system>.

AUTHOR
    Nathan Glenn <garfieldnate@gmail.com>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2012 by Nathan Glenn.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes

Revision history for Soar-Production-Parser
 
1.122860
	Released: 2012-10-12 04:31:32 UTC
	Merged stateImpCond and condsForOneId
	Removed rogue print statement

1.122830
	Released: 2012-10-09 19:41:36 UTC
	Require Perl 5.10.
	
1.122820
	Released: 2012-10-08 23:19:41 UTC
	Doc improvements and POD format fixes.

1.122800
	Released: 2012-10-06 22:47:02 UTC

t/TestSoarProdParser.pm
#
# This file is part of Soar-Production-Parser
#
# This software is copyright (c) 2012 by Nathan Glenn.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package t::TestSoarProdParser;
use Test::Base -Base;
use Soar::Production::Parser;

package t::TestSoarProdParser::Filter;
use Test::Base::Filter -base;
use Soar::Production::Parser;
use Data::Dumper;

my $parser = Soar::Production::Parser->new();

#return the parse of the text
sub parse { 
	# print "parse " . $_[0] . "\n"; 
	$parser->parse_text($_[0]);
}

#return 0 if the parse was unsuccessful, 1 if it was.	
sub parse_success {
	return 1 if defined $parser->parse_text($_[0]);
	return 0;
}

#an argument like LHS,stateImpCond,attrValueTests,0 will return
# $root->{LHS}->{stateImpCond}->{attrValueTests}->[0]
sub dive {
	my ($root) = shift @_;
	my @path = split ',', Test::Base::Filter::current_arguments();
	use Data::Diver qw (Dive);
	return Dive($root,@path);
}

#returns the number of productions found in a string
sub split_prods {
	my ($prods) = shift @_;
	my $productions = $parser->productions(text => $prods, parse => 0);
	# print $productions;
	return scalar @$productions;
}

#returns productions found in a string
sub get_prods {
	my ($prods) = shift @_;
	# print 'Getting prods from ' . $prods . "\n";
	my $args = Test::Base::Filter::current_arguments();
	my $productions = $parser->productions(text => $prods, parse => 0);
	# print "found $#$productions productions\n";
	return $productions
		unless defined $args;
		
	return $$productions[$args];
}

sub no_comment {
	my ($text) = @_;
	$text = $parser->no_comment($text);
	# print "returning: \n$text\n";
	return $text;
}

1;


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Data::Diver" => 0,
    "File::Find" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Module::Build" => "0.3601",
    "Test::Base" => 0,
    "Test::Base::Filter" => 0,
    "Test::Deep" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.88",
    "Test::Warn" => 0,
    "utf8" => 0
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "Parse Soar productions\r",
  "dist_author" => [
    "Nathan Glenn <garfieldnate\@gmail.com>"
  ],
  "dist_name" => "Soar-Production-Parser",
  "dist_version" => "1.122860",
  "license" => "perl",
  "module_name" => "Soar::Production::Parser",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "Parse::RecDescent" => 0,
    "perl" => "5.010",
    "strict" => 0,
    "warnings" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Soar/Production/Parser.pm
#
# This file is part of Soar-Production-Parser
#
# This software is copyright (c) 2012 by Nathan Glenn.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Soar::Production::Parser;
# ABSTRACT: Parse Soar productions

use strict;
use warnings;
#needed for advanced regex expressions
use 5.010;

our $VERSION = '1.122860'; # VERSION

use Soar::Production::Parser::PRDGrammar;
use Parse::RecDescent;
use Carp;
use Data::Dumper;

#a regular expression to split text into productions
my $splitter = qr/
	(sp\s+				#start with 'sp'	
	\{					#opening brace
	  (					#save to $2
		 (?: 			#either
			\{ (?-1) \}	#more braces and recurse
			|			#or
			(?:			#group
				[^|{}]++	#not bar or braces
				|   		#or
				\| 			#a bar
				(?:			#group
					[^\\|]++	#no slashes or bars, no backtracking
					|			#or
					\\.			#slash anything
				)*+ 
				\|
			)++				#one or more, no backtracking
		 )*				#0 or more times
	  )					#end $2
	\}					#ending brace
	)					#end $1
/x;

__PACKAGE__->new->_run(shift) unless caller;

sub _run {
  my ($soarParser, $filePath) = @_;
  print Dumper($soarParser->productions(file => $filePath, parse => 1) );
  return;
}

sub new {
  my ($class) = @_;
  my $soarParser = bless {}, $class;
  $soarParser->_init;
  return $soarParser;
}

sub _init {
  my ($soarParser) = @_;
    $soarParser->{parser} = Parse::RecDescent->new($Soar::Production::Parser::PRDGrammar::GRAMMAR);
	
	#if you wish to debug the grammar, try turning on traces by uncommenting the following lines:
	# $::RD_TRACE = 1;
	# $::RD_HINT = 1;
	return;
}


sub productions {## no critic RequireArgUnpacking
    my ($soarParser) = shift;
	my %args = (
		parse	=> 0,
		text	=> undef,
		file	=> undef,
		@_
	);
	defined $args{text} or defined $args{file}
		or croak 'Must specify parameter \'file\' or \'text\' to extract productions.';
	
	if($args{text}){
		return $soarParser->_productions_from_text($args{text}, $args{parse});
	}
	if($args{file}){
		# print "$args{file}\n";
		return $soarParser->_productions_from_file($args{file}, $args{parse});
	}
}

sub _productions_from_text {
    my ( $soarParser, $text, $parse) = @_;
	
	#remove comments
	# print 'Getting prods from ' . $text . "\n";
	$text = $soarParser->no_comment($text);
	# print "no_comments = $text";
	my $productions = _split_text(\$text);
	return $productions
		unless($parse);
	
	return $soarParser->get_parses($productions);
}

sub _productions_from_file {
    my ( $soarParser, $file, $parse) = @_;
    my $text = _readFile($file);
	
	my $productions = _split_text($text);
	return $productions
		unless($parse);
	
	return $soarParser->get_parses($productions);
}

#split text reference into production
sub _split_text {
	my ($text) = @_;
	#split the text into productions by looking for 'sp { ... }'
	my @productions;
	while($$text =~ /$splitter/g){
		# print "found production: $1";
		push @productions, $1;
	}
	return \@productions;
}

sub parse_text {
    my ( $soarParser, $text ) = @_;
    croak 'no text to parse!'
      unless defined $text;
    # $soarParser->{input} = \$text;
    return $soarParser->{parser}->parse($text);
}

#expects array reference holding production text
sub get_parses {
	my ($soarParser, $productions) = @_;
	my @parses;
	for(@$productions){
		# print STDERR $_;
		push @parses, $soarParser->{parser}->parse($_);
	}
	return \@parses;
}

#argument should be an opened file handle
#returns string pointer to text
sub _readFile {
    my ($file) = @_;
    open my $fh, '<', $file
		or croak "Couldn't open $file";
		
    my $text = q();
    $text .= $_ while (<$fh>);
	
	close $fh;
	$text = no_comment(undef,$text);
    return \$text;
}

#remove Soar comments from a line.
sub no_comment {
	#s/#(?=[^|]*$).*$//; #nope
	#s/#.*//g; #nope
	my ($parser, $text) = @_;
	$text =~ s/
			(			#save in $1
				\|			#literal bar
				(?:			#group
					\\[|]		#or an escaped bar
					|
					[^|]		#not a literal bar
				)*		#zero or more of previous group
				\|			#literal bar
			)			#end save
			|			#or
			(?:;\s*)?			# optional semicolon
			\#			# pound character
			.*			#followed by anything
		/			#replace with
			$1||''	# $1 or nothing (the quote if there was one; 
					# no quote will simply remove matching comment)
		/xeg;
	# print "returning $text";
	return $text;
}
1;

__END__

=pod

=head1 NAME

Soar::Production::Parser - Parse Soar productions

=head1 VERSION

version 1.122860

=head1 SYNOPSIS

  use Soar::Production::Parser;
  use Data::Dumper;
  my $parser = Soar::Production::Parser->new;
  my @trees=$parser->parse_file("foo.soar");
  print Dumper(\@trees);

=head1 DESCRIPTION

Soar is a cognitive modeling architecture for programming and experimenting with intelligent agents. Soar is programmed using productions that look like this:

	sp{name
		(state <s>)
		-->
		(<s> ^foo bar)
	}

The preceding production matches any state and adds an element named "foo" with the value "bar" to it. Productions can get much more complicated than that.
This module can be used to parse these productions. Underlyingly, a Parse::RecDescent grammar is used to convert a production into a parse tree. 
There are also methods for extracting all of the productions from a file string, and to remove comments (not that I think you'll ever want to do that!). 

=head1 NAME

Soar::Production::Parser - Perl extension for parsing angst grammar files

=head1 METHODS

=head2 C<new>

Creates a new parser.

=head2 C<parse_text>

Argument: the text of a single Soar production.
Returns: a parse tree for the given production.

=head2 C<get_parses>

Argument: Reference to array containing text for individual productions.
Return: Reference to an array containing parse trees for each of the productions in the input array reference.

=head2 C<no_comment>

Argument: Text which contains Soar productions or commands
Return: Same text, but with all comments removed. Comments are indicated with a # (pound), optionally preceded by a ; (semicolon) and whitespace.

=head2 C<productions>

This method extracts productions from a given text. It returns a reference to an array containing either the text of each of the productions, or a parse tree for each of them. Note that all comments are removed as a preprocessing step to detecting and extracting productions. It takes a set of named arguments:
'file'- the name of a file to read.
'text'- the text to split.
'parse'- set to true if the return value should be an array of parse trees for the extracted productions; otherwise an array containing the production text will be returned.
For example, if you would like to extract all of the productions from a file and print their parse trees, you could do this:

    use Soar::Production::Parser;
	use Data::Dumper;
	
	my $file = shift;
	my $parser = Soar::Production::Parser->new();
	my $parses = $parser->productions(
		file => $file,
		parse => 1
	);
	
	for my $prod(@$productions){
		print Dumper($prod);
	}

=head1 SEE ALSO

The documentation for Soar is located at L<https://code.google.com/p/soar/>.
You may also be interested in what a production system is, since this module parses Soar productions: L<http://en.wikipedia.org/wiki/Production_system>.

=head1 AUTHOR

Nathan Glenn <garfieldnate@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Nathan Glenn.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Soar/Production/Parser/PRDGrammar.pm
#
# This file is part of Soar-Production-Parser
#
# This software is copyright (c) 2012 by Nathan Glenn.
#
# This is free software; you can redistribute it and/or modify it under
# the same terms as the Perl 5 programming language system itself.
#
package Soar::Production::Parser::PRDGrammar;
# ABSTRACT: P::RD grammar for Soar productions

use strict;
use warnings;

our $VERSION = '1.122860'; # VERSION

#this grammar will return a parse tree of a production
our $GRAMMAR = <<'EOGRAMMAR';
	<nocheck>
	parse:
		<skip: qr{(?xs:
          (?: \s+						# Whitespace
          |   (?:;\s*)?\# [^\n]* \n?	# End of line comment
          )
		)*}> production /\Z/
		{
			$item[2]
		}
	#future work: can be sp or tp (normal or template production)
	production: /sp/ "{" beginning LHS "-->" RHS "}"
		{
			my %return;
			@return{ keys %{$item[3]}} = values %{$item[3]};
			$return{LHS} = $item{LHS};
			$return{RHS} = $item{RHS}->{rhsActions};
			\%return;
		}
	beginning: prodname documentation(?) flag(s?)
		{
			{name => $item[1], doc => $item[2] ? $item[2][0] : undef, flags => $item[3]}
		}
	prodname: /[\dA-Za-z][\dA-Za-z\$%&*=><?_\/\@:-]*/
	
	#documentation can span many lines
	documentation: '"' <commit> /[^"]*/ms '"'
		{
			$item[3]
		}
	flag: ':' <commit> /o-support|i-support|chunk|default|interrupt|template/
	LHS: cond(s)
		{
			{ conditions		=> $item[1] }
		}
	condType: "state" | "impasse"
	cond: 
		positiveCond 
		{
			{ negative => 'no', condition => $item{positiveCond} }
		}
		| negativeCond
		{
			{ negative => 'yes', condition => $item{negativeCond} }
		}
	negativeCond: "-" positiveCond
	positiveCond: condsForOneId
		{
			$item{condsForOneId};
		}
		| "{" <commit> cond(s) "}"
		{ 
			{ 'conjunction' => $item[3] }
		}
	condsForOneId: "(" <commit> condType(?) idTest(?) attrValueTests(s?) ")"
		#only a state_imp_cond can be missing an idTest or attrValueTests
		<reject: do { 
			not defined $item[3] and (
				not defined $item[4] or $#{$item[5]} == -1 
			)
		} >
		{
			{
				condType 	=> ($#{$item[3]} != -1 ? $item[3][0] : undef),
				idTest 		=> ($#{$item[4]} != -1 ? $item[4][0]->{test} : undef),
				attrValueTests =>	$item[5],
			}
		}
	idTest: test
		{
			{ test => $item{test} }
		}
	attrValueTests: /-?/ attTest valueTest(s?)
		{
			{
				negative 	=> ($item[1] ? 'yes' : 'no'),
				attrs 		=> $item[2],
				values		=> $item[3],
			}
		}
	attTest: "^" <commit> test(s /\./)
	valueTest: test /\+?/
		{
			{
				test 	=> $item{test},
				'+'		=> ($item[2] ? 'yes' : 'no'),
			}
		}
		| condsForOneId /\+?/
		{
			{
				conds	=> $item{condsForOneId},
				'+'		=> ($item[2] ? 'yes' : 'no'),
			}
		}
	test:
		conjunctiveTest
		{
			{ conjunctiveTest => $item{conjunctiveTest} }
		}
		| simpleTest
		{
			{ simpleTest => $item{simpleTest} }
		}
	conjunctiveTest: "\{" <commit> simpleTest(s) "\}"
		{
			$item[3]
		}
	simpleTest: 
		disjunctionTest
		{
			{ disjunctionTest => $item{disjunctionTest} }
		}
		| relationalTest
		{
			{ relationTest => $item{relationalTest} }
		}
		| singleTest
		{
			$item{singleTest}
		}
	disjunctionTest: /<<(?=\s)/ <commit> constant(s) />>/ #don't have to worry about look for whitespace on second one; if no space is there, the parser will think it's a string and fail.
	{
		$item[3]
	}
	relationalTest: relation singleTest #note that I removed a (?) from relation, and added singleTest to simpleTest
		{
			{ relation => ($item{relation} || undef), test => $item{singleTest} }
		}
	# /<(?=\s)/ ensures we don't match the beginning of a variable
	relation:  "<=>" | "<>" | "<=" | ">=" | ">" | /<(?=\s)/ | "=" 
	
	singleTest: 
		variable
		{
			$item{variable}
		}
		| constant
		{
			$item{constant}
		}
	
	#change skip so we can't have 
	variable: /<[A-Za-z0-9\$%&*+\/:=?_<>-]+(?<!>)>/
		{
			$item[1] =~ s/^<(.*)>$/$1/;
			{variable => $item[1] }
		}
	
	RHS: rhsAction(s?)
		{
			{ rhsActions => $item[1] }
		}
		
	rhsAction:
		funcCall
		{
			{ funcCall => $item{funcCall} }
		}
		| "(" variable attrValueMake(s) ")"
		{
			{ variable => $item{variable}->{variable}, attrValueMake => $item[3] }
		}
		
	funcCall: "(" funcName <commit> rhsValue(s?) ")"
		{
			{ function => $item{funcName}, args => $item[4] }
		}
	funcName: "+" | "-" | "*" | "/" | symConstant
	rhsValue: variable | constant | "(crlf)" | funcCall
	attrValueMake: <leftop: attr "." variableOrSymConstant> valueMake(s)
		{
			{ attr => $item[1], valueMake => $item[2] }
		}
	attr: "^" <commit> variableOrSymConstant
	variableOrSymConstant: variable | symConstant
	valueMake: rhsValue preferenceSpecifier(s?) 
		{
			#add an acceptable preference if no preference is specified
			my $preferences = $item[2];
			if($#$preferences == -1){
				$preferences = [{
                                  'value' => '+',
                                  'type' => 'unary'
                                }];
			}
			{ rhsValue => $item{rhsValue},  preferences => $preferences }
		}
		
	preferenceSpecifier: 
		unaryOrBinaryPreference rhsValue comma(?)
		{
			{ type => 'binary', value => $item{unaryOrBinaryPreference}, compareTo => $item{rhsValue} }
		}
		| unaryPreference comma(?)
		{
			{ type => 'unary', value => $item{unaryPreference} };
		}
		
	comma: ","
		
	unaryPreference: "+" | "-" | "!" | "~" | "@" | unaryOrBinaryPreference
	
	#negative lookahead necessary to prevent matching <xx> as two specifiers and a constant
	unaryOrBinaryPreference: ">" | ...!variable  "<" | "=" | "&"
	
	#put float and int first, since symConstant can technically match the same values.
	constant: 
		floatConstant
		{
			{ constant => $item{floatConstant}, type => 'float' }
		}
		| intConstant
		{
			{ constant => $item{intConstant}, type => 'int' }
		}
		| symConstant 
		{
			{ constant => $item{symConstant}, type => 'sym' }
		}
	symConstant: string { $item{string} } | quoted { $item{quoted} }
	string: /[A-Za-z0-9\$%&*+\/:=?_><-]+/ 
		<reject: do{ $item[1] =~ /^<.*>$/} > #reject if we've actually found a variable
		<reject: do{ 
			$item[1] =~ /^ [+!~><=-]+ $/x and 
			$item[1] !~ /^ (?: >< | [<>]{3,}) $/x 
			
			} 
		> #reject if the name contains only preference characters
		<defer: {
			# "<x" or "x>" look like typos. Could have missed a pointy brace.
			if( $item[1] =~ /^<.*|.*>$/ ){
				use Carp;
				carp "Suspicious string constant: \"$item[1]\". Did you mean to use a variable or disjunction?";
			}
		}>
		{
			# "<x" or "x>" look like typos. Could have missed a pointy brace. Convert to quoted like Soar does.
			if( $item[1] =~ /^<.*|.*>$/ ){
				$return = { type => 'quoted', value => $item[1] };
			}else{
				$return = { type => 'string', value => $item[1] };
			}
		}
	#TODO: note that in Soar, || is ignored and treated like <any>.
	quoted: /\|(?:\\[|]|[^|])*\|/
		{
			#remove leading and trailing vertical bar
			$item[1] =~ s{^\|}{};
			$item[1] =~ s{\|$}{};
			
			#unescape other vertical bars
			$item[1] =~ s{\\\|}{|}g;
			
			{ type => 'quoted', value => $item[1] }
		}
	intConstant: /-?[0-9]+/
	floatConstant: 
		scientific { $item{scientific} }
		| normal { $item{normal} }
	
	#strangely enough, the section after the period is optional; '1.' is legal.
	normal: /^[-+]?[0-9]*\.[0-9]*/
	scientific: /[+-]?[0-9]\.[0-9]+[eE][-+]?[0-9]+/
EOGRAMMAR

1;

__END__

=pod

=head1 NAME

Soar::Production::Parser::PRDGrammar - P::RD grammar for Soar productions

=head1 VERSION

version 1.122860

=head1 SYNOPSIS

  use Soar::Production::Parser::PRDGrammar;
  use Parse::RecDescent;
  
  my $parser = Parse::RecDescent->new($Soar::Production::Parser::PRDGrammar::GRAMMAR);
  $parser->parse($some_text);

=head1 DESCRIPTION

This module holds one string: a C<Parse::RecDescent> grammar for parsing Soar productions.

=head1 NAME

Soar::Production::Parser::PRDGrammar - C<Parse::RecDescent> grammar for parsing Soar productions.

=head1 BNF Grammar from the Soar Manual

The following is a description of a grammar describing Soar productions, from the Soar manual.
Note also that a /(;\s*)?\#/ (pound sign optionally preceded by semi-colon and white space) indicates an end-of-line comment.
Soar does not have multi-line comments.

=head2 Grammar of Soar productions

A grammar for Soar productions is:

	<soar-production> ::= sp "{" <production-name> [<documentation>] [<flags>]
	<condition-side> --> <action-side> "}"
	<documentation> ::= """ [<string>] """
	<flags> ::= ":" (o-support | i-support | chunk | default)

=head3 Grammar for Condition Side

I have modified the grammar below from the original to account for the fact that an instance of <state-imp-cond> is not required to start the LHS, and may actually occur anywhere inside of it. 

	Below is a grammar for the condition sides of productions:
	<condition-side> ::= <cond>+
	<cond> ::= <positive_cond> | "-" <positive_cond>
	<positive_cond> ::= <state-imp-cond> | <conds_for_one_id> | "{" <cond>+ "}"
	<state-imp-cond> ::= "(" (state | impasse) [<id_test>] <attr_value_tests>+ ")"
	<conds_for_one_id> ::= "(" [(state|impasse)] <id_test> <attr_value_tests>+ ")"
	<id_test> ::= <test>
	<attr_value_tests> ::= ["-"] "^" <attr_test> ("." <attr_test>)* <value_test>*
	<attr_test> ::= <test>
	<value_test> ::= <test> ["+"] | <conds_for_one_id> ["+"]
	<test> ::= <conjunctive_test> | <simple_test>
	<conjunctive_test> ::= "{" <simple_test>+ "}"
	<simple_test> ::= <disjunction_test> | <relational_test>
	<disjunction_test> ::= "<<" <constant>+ ">>"
	<relational_test> ::= [<relation>] <single_test>
	<relation> ::= "<>" | "<" | ">" | "<=" | ">=" | "=" | "<=>"
	<single_test> ::= <variable> | <constant>
	<variable> ::= "<" <sym_constant> ">"
	<constant> ::= <sym_constant> | <int_constant> | <float_constant>

=head4 Notes on the Condition Side

In an <id_test>, only a <variable> may be used in a <single_test>.

=head3 Grammar for Action Side

Below is a grammar for the action sides of productions:

	<rhs> ::= <rhs_action>*
	<rhs_action> ::= "(" <variable> <attr_value_make>+ ")" | <func_call>
	<func_call> ::= "(" <func_name> <rhs_value>* ")"
	<func_name> ::= <sym_constant> | "+" | "-" | "*" | "/"
	<rhs_value> ::= <constant> | <func_call> | <variable>
	<attr_value_make> ::= "^" <variable_or_sym_constant> ("." <variable_or_sym_constant>)* <value_make>+
	<variable_or_sym_constant> ::= <variable> | <sym_constant>
	<value_make> ::= <rhs_value> <preference_specifier>*
	<preference-specifier> ::= <unary-preference> [","]
		| <unary-or-binary-preference> [","]
		| <unary-or-binary-preference> <rhs_value> [","]
	<unary-pref> ::= "+" | "-" | "!" | "~" | "@"
	<unary-or-binary-pref> ::= ">" | "=" | "<" | "&"

=head1 CAVEATS

This grammar does not check for the semantic correctness of a production. For example, the following parses just fine:

	sp {foo
		(state)
	-->
		(<s> ^foo <a>)
	}
Whereas the real Soar parser gives us an error, saying that <lt>s<gt> is not connected to the LHS in any way.

=head1 BUGS

This may belong in the realm of semantic verification, but for now it is considered a known bug. In Soar, numbers cannot be used to name a working memory element unless they are quoted. This is not yet implemented in this grammar, and thus the following is accepted, even though Soar would reject it:

	sp {bad-numbers
		(state <s> ^string.2.string)
	-->
	}

=head1 SEE ALSO

The documentation for Soar is located at L<https://code.google.com/p/soar/>

=head1 AUTHOR

Nathan Glenn <garfieldnate@gmail.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Nathan Glenn.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Parse Soar productions\r",
  "AUTHOR" => "Nathan Glenn <garfieldnate\@gmail.com>",
  "BUILD_REQUIRES" => {
    "Data::Diver" => 0,
    "File::Find" => 0,
    "File::Spec::Functions" => 0,
    "File::Temp" => 0,
    "FindBin" => 0,
    "Module::Build" => "0.3601",
    "Test::Base" => 0,
    "Test::Base::Filter" => 0,
    "Test::Deep" => 0,
    "Test::Exception" => 0,
    "Test::More" => "0.88",
    "Test::Warn" => 0,
    "utf8" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Soar-Production-Parser",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Soar::Production::Parser",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Data::Dumper" => 0,
    "Parse::RecDescent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "1.122860",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



